opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/reminder/ReminderHelper.java

```diff
@@ -0,0 +1,133 @@
+package de.geeksfactory.opacclient.reminder;
+
+import org.joda.time.LocalDate;
+import org.joda.time.format.DateTimeFormat;
+
+import android.app.AlarmManager;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import de.geeksfactory.opacclient.BuildConfig;
+import de.geeksfactory.opacclient.OpacClient;
+import de.geeksfactory.opacclient.objects.LentItem;
+import de.geeksfactory.opacclient.storage.AccountDataSource;
+
+public class ReminderHelper {
+    private OpacClient app;
+    private SharedPreferences sp;
+
+    public ReminderHelper(OpacClient app) {
+        this.app = app;
+        sp = PreferenceManager.getDefaultSharedPreferences(app);
+    }
+
+    /**
+     * Save alarms for expiring media to the DB and schedule them using {@link
+     * android.app.AlarmManager}
+     */
+    public void generateAlarms() {
+        int warning = Integer.parseInt(sp.getString("notification_warning", "3"));
+
+        AccountDataSource data = new AccountDataSource(app);
+        data.open();
+        List<LentItem> items = data.getAllLentItems();
+
+        // Sort lent items by deadline
+        Map<LocalDate, List<Long>> arrangedIds = new HashMap<>();
+        for (LentItem item : items) {
+            LocalDate deadline = item.getDeadline();
+            if (!arrangedIds.containsKey(deadline)) {
+                arrangedIds.put(deadline, new ArrayList<Long>());
+            }
+            arrangedIds.get(deadline).add(item.getDbId());
+        }
+
+        // Remove alarms with no corresponding media
+        for (Alarm alarm : data.getAllAlarms()) {
+            if (!arrangedIds.containsKey(alarm.deadline)) {
+                data.removeAlarm(alarm);
+            }
+        }
+
+        // Find and add/update corresponding alarms for current lent media
+        for (Map.Entry<LocalDate, List<Long>> entry : arrangedIds.entrySet()) {
+            LocalDate deadline = entry.getKey();
+            long[] media = toArray(entry.getValue());
```

Comment:
> Regarding the comma-separated strings in the database: Is there even a better way?

The "normalized" way would be to have three tables: ``alarm``, ``lent_item``, ``lent_item_on_alarm``. But I think it is overkill here as we do not need the features of relational databases.

> But it would cause the expired items to be displayed in the notification in a different order every time the notification is shown

Well, we could sort it at display time. 

> and if we consequently also changed the List<LentItem> and List<ReservedItem> in the AccountData class to Sets, that would definitely be confusing because the order would change every time the user re-opens or refreshes the account view.

You're right about that, I take my suggestion back. 

On the other hand, this "convert to array, than use Array.equals" compares order-sensitive. Should we use a Set instead of an array just here?